src/arcsf/data/tofu.py

```diff
@@ -0,0 +1,193 @@
+        if isinstance(out[0], list):
+            return _flatten_list_of_lists(out)
+        return out
+
+
+def load_tofu(
+    granularity: str,
+    stratified: bool,
+    forget_random: bool,
+    forgotten_author_fraction: float,
+    forgotten_fact_fraction: float,
+    random_seed: int,
+    debug=False,
+) -> tuple[Dataset, Dataset, dict] | tuple[Dataset, Dataset]:
+    """
+    Loads TOFU dataset given different flags for retain--forget split.
+    Args:
+        - granularity: level at which forgetting takes place (author vs question)
+        - stratified: if forgetting questions restrain to specific authors?
+        - forget_random: is forgetting happening randomly within constraints?
+        - forgotten_author_fraction: fraction of authors from which to forget questions
+        - forgotten_fact_fraction: fraction of questions to randomly forget
+            - if stratified == True represents fraction of Qs in author
+            - if stratified == False represents fraction of total Qs
+        - random_seed: seed for reproducibility
+        - debug: returns dictionary containing meta_data if being used during testing
+    Returns:
+        - Two datasets with forget and retain sets
+            as well as a debugging dictionary (optional)
+    """
+
+    all_data = load_dataset("locuslab/TOFU", "full")["train"]
+
+    num_authors = 200  # hard coding author count for now
+    q_per_author = 20  # hard coding author question count for now
+
+    q_indices = [q for q in range(num_authors * q_per_author)]
+    a_indices = [math.floor(q / q_per_author) for q in q_indices]
+
+    all_data = all_data.add_column("question_index", q_indices)
+    all_data = all_data.add_column("author_index", a_indices)
+
+    # if author, then the number of questions to forget (per author) is all of them
+    if granularity == "author":
+        num_forget = q_per_author
+    # else: it depends on other arguments
+    elif granularity == "question":
+        if stratified:
+            # non-random stratification: hard code 4 questions for now
+            if not forget_random:
+                num_forget = int(4)
+            # random then it depends on q_per_author and forgotten_fact_fraction
+            elif forget_random:
+                num_forget = int(q_per_author * forgotten_fact_fraction)
+        # non stratified random: forgotten_fact_fraction applied across entire dataset
+        elif not stratified:
+            num_forget = int(num_authors * q_per_author * forgotten_fact_fraction)
+
+    debug_dict = {
+        "author_count": num_authors,
+        "q_per_author": q_per_author,
+        "num_forget": num_forget,
+    }
+
+    # Get full list of indices
+    all_indices = list(range(all_data.num_rows))
+
+    # Get full list of authors
+    authors = list(range(num_authors))
+
+    # Conditionally create full list of questions
+    if granularity == "author" or not forget_random:
+        forget_questions = list(range(num_forget))
+
```

Comment:
this doesn't seem to be used again until line 172, can it be moved closer to there to help understand what it's for?